usr/sbin/p11sak/p11sak.c

```diff
@@ -3510,133 +3664,167 @@ static CK_RV get_key_infos(CK_OBJECT_HANDLE key, CK_OBJECT_CLASS *class,
     }
 
-    switch (class_val) {
-    case CKO_PUBLIC_KEY:
-    case CKO_PRIVATE_KEY:
-    case CKO_SECRET_KEY:
-        break;
-    default:
-        free(attrs[0].pValue);
-        return CKR_KEY_NEEDED;
+    attr.pValue = calloc(attr.ulValueLen + 1, 1);
+    if (attr.pValue == NULL) {
+        warnx("Failed to allocate memory for label attribute");
+        return CKR_HOST_MEMORY;
     }
 
-    for (i = 0; i < num_attrs; i++) {
-        if (attrs[i].ulValueLen == CK_UNAVAILABLE_INFORMATION) {
-            warnx("Attribute %s is not available in key object",
-                  p11_get_cka(attrs[i].type));
-            free(attrs[0].pValue);
-            return CKR_TEMPLATE_INCOMPLETE;
-        }
+    rv = pkcs11_funcs->C_GetAttributeValue(pkcs11_session, obj, &attr, 1);
+    if (rv != CKR_OK) {
+        warnx("Failed to get CKA_LABEL attribute: C_GetAttributeValue: 0x%lX: %s",
+              rv, p11_get_ckr(rv));
+        free(attr.pValue);
+        return rv;
     }
 
-    if (class != NULL)
-        *class = class_val;
-    if (ktype != NULL)
-        *ktype = ktype_val;
+done:
+    *label_value = attr.pValue;
 
-    keytype_val = find_keytype(ktype_val);
-    if (keytype_val == NULL) {
-        warnx("Key object \"%s\" has an unsupported key type: %lu",
-              (char *)attrs[0].pValue, ktype_val);
-        free(attrs[0].pValue);
-        return CKR_KEY_TYPE_INCONSISTENT;
-    }
+    return CKR_OK;
+}
+
+static CK_RV get_obj_infos(CK_OBJECT_HANDLE obj, CK_OBJECT_CLASS *class,
+                           CK_ULONG *otype, CK_ULONG *keysize,
+                           char** label, char** typestr,
+                           const struct p11sak_objtype **objtype,
+                           char **common_name)
+{
+    CK_RV rc;
+    CK_OBJECT_CLASS class_val = 0;
+    CK_ULONG otype_val = 0;
+    const struct p11sak_objtype *objtype_val;
+    CK_ULONG keysize_val = 0;
+    char *label_val = NULL;
+    char *common_name_val = NULL;
 
-    if (keytype != NULL)
-        *keytype = keytype_val;
+    rc = get_label_value(obj, &label_val);
+    if (rc != CKR_OK)
+        return rc;
 
-    if (keytype_val->keysize_attr != (CK_ATTRIBUTE_TYPE)-1) {
-        keysize_attr.type = keytype_val->keysize_attr;
-        if (!keytype_val->keysize_attr_value_len) {
-            keysize_attr.ulValueLen = sizeof(keysize_val);
-            keysize_attr.pValue = &keysize_val;
-        } else {
-            /* Query attribute length only */
-            keysize_attr.ulValueLen = 0;
-            keysize_attr.pValue = NULL;
+    rc = get_class_and_type_values(obj, label_val, &class_val, &otype_val);
+    if (rc != CKR_OK)
+        return rc;
+
+    switch (class_val) {
+    case CKO_SECRET_KEY:
+    case CKO_PUBLIC_KEY:
+    case CKO_PRIVATE_KEY:
+        objtype_val = find_keytype(otype_val);
+        if (objtype_val == NULL) {
+            warnx("%s object \"%s\" has an unsupported %s type: %lu",
+                  objtype_val->obj_liststr, label_val,
```

Comment:
You can not reference objtype_val to print the message since objtype_val is NULL (2 times) in this statement.

Reply:
Yep, that's right ... changed msg ...